controller/src/main/java/io/pravega/controller/store/stream/ZkInt96Counter.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    CompletableFuture<Int96> getNextCounter() {
+        CompletableFuture<Int96> future;
+        synchronized (lock) {
+            Int96 next = counter.incrementAndGet();
+            if (next.compareTo(limit.get()) > 0) {
+                // ignore the counter value and after refreshing call getNextCounter
+                future = refreshRangeIfNeeded().thenCompose(x -> getNextCounter());
+            } else {
+                future = CompletableFuture.completedFuture(next);
+            }
+        }
+        return future;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> refreshRangeIfNeeded() {
+        CompletableFuture<Void> refreshFuture;
+        synchronized (lock) {
+            // Ensure that only one background refresh is happening. For this we will reference the future in refreshFutureRef
+            // If reference future ref is not null, we will return the reference to that future.
+            // It is set to null when refresh completes.
+            refreshFuture = this.refreshFutureRef;
+            if (this.refreshFutureRef == null) {
+                // no ongoing refresh, check if refresh is still needed
+                if (counter.get().compareTo(limit.get()) >= 0) {
+                    log.info("Refreshing counter range. Current counter is {}. Current limit is {}", counter.get(), limit.get());
+
+                    // Need to refresh counter and limit. Start a new refresh future. We are under lock so no other
+                    // concurrent thread can start the refresh future.
+                    refreshFutureRef = getRefreshFuture()
+                            .exceptionally(e -> {
+                                // if any exception is thrown here, we would want to reset refresh future so that it can be retried.
+                                synchronized (lock) {
+                                    refreshFutureRef = null;
+                                }
+                                log.warn("Exception thrown while trying to refresh transaction counter range", e);
+                                throw new CompletionException(e);
+                            });
+                    // Note: refreshFutureRef is reset to null under the lock, and since we have the lock in this thread
+                    // until we release it, refresh future ref cannot be reset to null. So we will always return a non-null
+                    // future from here.
+                    refreshFuture = refreshFutureRef;
+                } else {
+                    // nothing to do
+                    refreshFuture = CompletableFuture.completedFuture(null);
+                }
+            }
+        }
+        return refreshFuture;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> getRefreshFuture() {
+        return storeHelper.createZNodeIfNotExist(COUNTER_PATH, Int96.ZERO.toBytes())
+                          .thenCompose(v -> storeHelper.getData(COUNTER_PATH, Int96::fromBytes)
```

Comment:
Remind me of one thing: are we using `chroot`? Otherwise we are writing this under the root node, which is not ideal.

Reply:
yes, we do use `namespace` while creating the curator client which makes the namespace as the root node. 
The namespace used is `pravega` so all these go under `/pravega/`